tests/src/test/spark330/scala/org/apache/spark/sql/RapidsTestsTrait.scala

```diff
@@ -0,0 +1,356 @@
+  }
+
+  def rapidsCheckExpression(expression: Expression, expected: Any, inputRow: InternalRow): Unit = {
+    val df = if (inputRow != EmptyRow && inputRow != InternalRow.empty) {
+      convertInternalRowToDataFrame(inputRow)
+    } else {
+      val schema = StructType(StructField("a", IntegerType, nullable = true) :: Nil)
+      val empData = Seq(Row(1))
+      _spark.createDataFrame(_spark.sparkContext.parallelize(empData), schema)
+    }
+    val resultDF = df.select(Column(expression))
+    val result = resultDF.collect()
+    TestStats.testUnitNumber = TestStats.testUnitNumber + 1
+    if (
+      checkDataTypeSupported(expression) &&
+      expression.children.forall(checkDataTypeSupported)
+    ) {
+      val projectTransformer = resultDF.queryExecution.executedPlan.collect {
+        case p: GpuProjectExec => p
+      }
+      if (projectTransformer.size == 1) {
+        TestStats.offloadRapidsUnitNumber += 1
+        logInfo("Offload to native backend in the test.\n")
+      } else {
+        logInfo("Not supported in native backend, fall back to vanilla spark in the test.\n")
+        shouldNotFallback()
+      }
+    } else {
+      logInfo("Has unsupported data type, fall back to vanilla spark.\n")
+      shouldNotFallback()
+    }
+
+    if (
+      !(checkResult(result.head.get(0), expected, expression.dataType, expression.nullable)
+        || checkResult(
+          CatalystTypeConverters.createToCatalystConverter(expression.dataType)(
+            result.head.get(0)
+          ), // decimal precision is wrong from value
+          CatalystTypeConverters.convertToCatalyst(expected),
+          expression.dataType,
+          expression.nullable
+        ))
+    ) {
+      val input = if (inputRow == EmptyRow) "" else s", input: $inputRow"
+      fail(
+        s"Incorrect evaluation: $expression, " +
+          s"actual: ${result.head.get(0)}, " +
+          s"expected: $expected$input")
+    }
+  }
+
+  def shouldNotFallback(): Unit = {
+    TestStats.offloadRapids = false
+
+    // TODO
+  }
+
+  def canConvertToDataFrame(inputRow: InternalRow): Boolean = {
+    if (inputRow == EmptyRow || inputRow == InternalRow.empty) {
+      return true
+    }
+    if (!inputRow.isInstanceOf[GenericInternalRow]) {
+      return false
+    }
+    val values = inputRow.asInstanceOf[GenericInternalRow].values
+    for (value <- values) {
+      value match {
+        case _: MapData => return false
+        case _: ArrayData => return false
+        case _: InternalRow => return false
+        case _ =>
+      }
+    }
+    true
+  }
+
+  def convertInternalRowToDataFrame(inputRow: InternalRow): DataFrame = {
+    val structFileSeq = new ArrayBuffer[StructField]()
+    val values = inputRow match {
+      case genericInternalRow: GenericInternalRow =>
+        genericInternalRow.values
+      case _ => throw new UnsupportedOperationException("Unsupported InternalRow.")
+    }
+    @nowarn
```

Comment:
Curious what warning is being suppressed here?  Typically we want to be very specific about the warning we're trying to suppress so we don't inadvertently suppress a critical warning that must be fixed.  If warning suppression is really needed here, it should use `@scala.annotation.nowarn` with selective warning text as done elsewhere in the code.